### PR TITLE
Publishing context manager.

### PR DESCRIPTION
**Pull request by jortel**
https://pulp.plan.io/issues/3295

I originally added `add_artifact()` and `add_metadata()` as defined by the issue but after writing them they did not seem to provide much value.  Perhaps more discussion is needed.

### CODE REVIEW

File: pulpcore/pulpcore/app/models/publication.py [hunk 1]
```diff
@@ -20,6 +42,68 @@ class Publication(Model):
 
     repository_version = models.ForeignKey('RepositoryVersion', on_delete=models.CASCADE)
 
+    complete = False  # placeholder
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/models/publication.py]: dont forget to make this a model.Boolean

File: pulpcore/pulpcore/app/models/publication.py [hunk 2]
```diff
@@ -20,6 +42,68 @@ class Publication(Model):
...
+
+        Args:
+            **kwargs (dict): Delete options.
+
+        Notes:
+            Deletes the Task.created_resource when incomplete.
+        """
+        with transaction.atomic():
+            if not self.complete:
+                CreatedResource.objects.delete(object_id=self.pk)
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/models/publication.py]: I think we should delete the CreatedResource even if the Publication was never marked complete.

File: pulpcore/pulpcore/app/models/publication.py [hunk 3]
```diff
@@ -1,17 +1,39 @@
...
 
-from pulpcore.app.models import Model, storage
+from . import storage
+from .base import Model
+from .task import CreatedResource
 
 
 class Publication(Model):
     """
+    A publication.
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/models/publication.py]: From the language guide: 
Publications refer to the metadata and arifacts of the content units in a repository version.

File: pulpcore/pulpcore/app/models/publication.py [hunk 4]
```diff
@@ -20,6 +42,68 @@ class Publication(Model):
...
+            repository_version (pulpcore.app.models.RepositoryVersion): The repository
+                version to be published.
+            publisher (pulpcore.app.models.Publisher): The publisher used
+                to create the publication.
+
+        Returns:
+            pulpcore.app.models.Publication: self
+
+        Notes:
+            The created publication with complete=False.
```
Comment (asmacdo) [inline pulpcore/pulpcore/app/models/publication.py]: Would you move this note to the returns line? Something like
Returns:
     A created Publication in an incomplete state.

Comment (jortel) [reply pulpcore/pulpcore/app/models/publication.py]: 👍

File: pulpcore/pulpcore/app/models/publication.py (hunk 1 above)
Comment (jortel) [reply pulpcore/pulpcore/app/models/publication.py]: The complete field will be added by #3299.  This just a placeholder to get it working that will be removed before merge.

File: pulpcore/pulpcore/app/models/publication.py (hunk 2 above)
Comment (jortel) [reply pulpcore/pulpcore/app/models/publication.py]: 👍

**Pull request merged**